web/src/compose_ui.js

```diff
@@ -550,6 +550,138 @@ export function format_text($textarea, type, inserted_content) {
         wrapSelection(field, syntax_start, syntax_end);
     };
 
+    const format_spoiler = () => {
+        let spoiler_syntax_start = "```spoiler \n";
+        const spoiler_syntax_start_without_break = "```spoiler ";
+        let spoiler_syntax_end = "\n```";
+
+        // Captures:
+        // ```spoiler |
+        // <text>
+        // ```
+        // (with | as placeholder for the cursor)
+        const is_cursor_at_header_position = () =>
+            range.length === 0 &&
+            text.slice(range.start - spoiler_syntax_start_without_break.length, range.start) ===
+                spoiler_syntax_start_without_break;
+
+        // Captures:
+        // ```spoiler <header>
+        // <text>
+        // ```
+        // with <text> selected
+        const is_content_selected = () =>
+            text.slice(range.end, range.end + spoiler_syntax_end.length) === spoiler_syntax_end &&
+            text[range.start - 1] === "\n" &&
+            text.lastIndexOf(spoiler_syntax_start_without_break, range.start - 1) ===
+                text.lastIndexOf("\n", range.start - 2) + 1;
+
+        if (is_cursor_at_header_position()) {
+            const new_range_end = text.indexOf(spoiler_syntax_end, range.start);
+            text =
+                text.slice(0, range.start - spoiler_syntax_start_without_break.length) +
+                text.slice(range.start + 1, new_range_end) +
+                text.slice(new_range_end + spoiler_syntax_end.length);
```

Comment:
I think supported by moving this just after the definition of `is_cursor_at_header_position`, I'd really like to see the comment defining what we're capturing combined with better variable names to make this readable... it's pretty inscrutable.

In this case, I think `text.slice(range.start + 1, new_range_end)` is supposed to be the selected text that we're saving, and I'm not entirely sure why `text.slice(new_range_end + spoiler_syntax_end.length);` is the final construction rather than something referencing `range.end`, for example.

Reply:
When the cursor is at the header (this particular case), `range.end` is simply the cursor position which is before all the content inside the spoiler, with 0 relation to the actual end of the spoiler block. The `new_range_end` is the position where the content inside the spoiler block ends, hence we use that here, to get all the text after the spoiler block.

I have refactored this function, removing the `else if`s, adding clearer comments, etc. Also changed this entire header section to support adding and removing headers in spoiler blocks while toggling using this button